llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -1406,6 +1406,46 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
   return nullptr;
 }
 
+/// Fold the following code sequence:
+/// \code
+///   %XEq = icmp eq i64 %X, %Z
+///   %YEq = icmp eq i64 %Y, %Z
+///   %either = select i1 %XEq, i1 true, i1 %YEq
+///   %both = select i1 %XEq, i1 %YEq, i1 false
+///   %cmp = icmp eq i64 %X, %Y
+///   %equal = select i1 %either, i1 %both, i1 %cmp
+/// \code
+///
+/// into:
+///   %equal = icmp eq i64 %X, %Y
+///
+/// Equivalently:
+///   (X==Z || Y==Z) ? (X==Z && Y==Z) : X==Y --> X==Y
+Instruction *InstCombinerImpl::foldSelectEqualityTest(SelectInst &Sel) {
+  Value *X, *Y, *Z, *XEq, *YEq;
+  Value *Either = Sel.getCondition(), *Both = Sel.getTrueValue(),
+        *Cmp = Sel.getFalseValue();
+
+  if (!match(Either, m_Select(m_Value(XEq), m_One(), m_Value(YEq))))
+    return nullptr;
+
+  if (!match(XEq,
+             m_c_SpecificICmp(ICmpInst::ICMP_EQ, m_Value(X), m_Value(Z))) ||
```

Comment:
The commutative here isn't doing anything since your just match any value on either side.

Reply:
This should look something like:
```
if(!match(XEq, m_SpecificICmp(ICMP_EQ, m_Value(X), m_Value(Z))
  return nullptr;
if(!match(m_c_SpecificICmp(ICMP_EQ, m_Value(Y), m_Specific(Z)))
  std::swap(X, Z);
if(!match(m_c_SpecificICmp(ICMP_EQ, m_Value(Y), m_Specific(Z)))
  return nullptr;
```

Reply:
Ah, this points out a case that my tests were missing: `icmp eq 0, %Y` got canonicalized to have the zero on the RHS, so I wasn't covering the case that needed `swap`. I've added a case that uses X, Y, Z to avoid this.